deviations/models.py

```diff
@@ -1,15 +1,72 @@
 from datetime import datetime, timedelta
-from typing import Optional
+from typing import TYPE_CHECKING, Generic, Iterable, Optional, TypeVar, Union
 
 from django.urls import reverse
 from django.db import models
 from django.utils.translation import gettext_lazy as _
 
 from exercise.exercise_models import BaseExercise
+from exercise.submission_models import Submission
 from userprofile.models import UserProfile
 from lib.models import UrlMixin
 
 
+TModel = TypeVar('TModel', bound='SubmissionRuleDeviation')
+class SubmissionRuleDeviationManager(models.Manager[TModel], Generic[TModel]):
+    max_order_by: str
+
+    def get_max_deviations(
+        self,
+        submitter: UserProfile,
+        exercises: Iterable[Union[BaseExercise, int]],
+    ) -> Iterable[TModel]:
+        """
+        Returns the maximum deviations for the given submitter in the given
+        exercises (one deviation per exercise is returned). The deviation may
+        be granted to the submitter directly, or to some other submitter in
+        their group.
+        """
+        deviations = (
+            self.filter(
+                models.Q(exercise__in=exercises)
+                & (
+                    # Check that the owner of the deviation is the user, or
+                    # some other user who has submitted the deviation's
+                    # exercise with the user.
+                    models.Q(submitter=submitter)
+                    | models.Exists(
+                        # Note the two 'submitters' filters.
+                        Submission.objects.filter(
+                            exercise=models.OuterRef('exercise'),
+                            submitters=models.OuterRef('submitter'),
+                        ).filter(
+                            submitters=submitter,
```

Comment:
The two filters ensure that `submitter` (the user given as a parameter) and `OuterRef('submitter')` (another user) are **both** among the submitters of the same submission. This is the only way I could figure out how this could be done in Django's ORM. I found it on some StackOverflow question. If you look at the SQL code generated by this query, it creates **two** joins between the submission and its submitters, and filters one of them by `submitter` and one by `OuterRef('submitter')`.

This is the same code that I came up with back when we were both trying to fix the issue where the wrong user's deviations were stored in the cache, it's just been moved from `CachedPoints` to here. But I don't blame you if you forgot this explanation (if I even explained it back then), it was one exhausting evening...

Reply:
Nice recap! :+1: